src/app/notebook/formatted-notebook.component.ts

```diff
@@ -0,0 +1,375 @@
+  @Input() baseUrl: string;
+  @ViewChild('notebookTarget') notebookTarget: ElementRef;
+  loading = true;
+  displayError = false;
+  private currentSubscription: Subscription = null;
+
+  ngOnChanges() {
+    this.retrieveAndFormatNotebook();
+  }
+
+  retrieveAndFormatNotebook() {
+    this.notebookTarget?.nativeElement.replaceChildren(); // Remove the current formatted notebook.
+    this.loading = true;
+    this.displayError = false;
+    // The next line cancels any previous request that is still in progess,
+    // because if we're here, the @Inputs have changed, and we're about to
+    // launch a new request to retrieve the corresponding notebook file,
+    // which we will display when it arrives.  The previous response is now
+    // irrelevant, and could actually cause problems if it arrives out of
+    // order.
+    // Assuming that Angular calls a component's `ngOnChanges()` directly
+    // after setting its @Input fields, a pleasant side effect of said
+    // cancellation is that when we handle a response, we know that @Input
+    // fields have the same values as when the corresponding request was
+    // started, which simplifies the code.  For example, the methods
+    // called by the handler to format the notebook can safely use the
+    // value of the `workflow` field to determine the notebook language.
+    this.currentSubscription?.unsubscribe();
+    this.currentSubscription = this.workflowsService
+      .getWorkflowVersionsSourcefiles(this.workflow.id, this.version.id, ['DOCKSTORE_JUPYTER'])
+      .pipe(
+        finalize(() => {
+          this.loading = false;
+        })
+      )
+      .subscribe(
+        (sourceFiles: SourceFile[]) => {
+          for (const sourceFile of sourceFiles) {
+            if (this.isPrimaryDescriptor(sourceFile.path)) {
+              try {
+                // Create an element containing the formatted notebook,
+                // and make it the single child of the template's '#notebookTarget' placeholder.
+                const notebookElement = this.createFormattedNotebookElement(sourceFile.content);
+                this.notebookTarget.nativeElement.replaceChildren(notebookElement);
+              } catch (e) {
+                console.log('Exception formatting notebook');
+                console.log(e.message);
+                this.displayError = true;
+                return;
+              }
+              this.displayError = false;
+              return;
+            }
+          }
+          this.displayError = true;
+        },
+        () => {
+          this.displayError = true;
+        }
+      );
+  }
+
+  isPrimaryDescriptor(path: string): boolean {
+    return path === this.version.workflow_path;
+  }
+
+  /**
+   * Convert the specified notebook to a DOM element that represents it.
+   */
+  createFormattedNotebookElement(notebook: string): any {
+    // Create the notebook container div.
+    const element: any = this.document.createElement('div');
```

Comment:
> Couldn't you have something like this in the HTML template:
> 
> ```
> <div class="notebook" *ngIf="<condition>" [innerHtml]="sanitizedHtml">
> <!--More stuff...-->
> </div>
> ```
> 
> I find this more readable than tracing through TypeScript code to see what HTML will be generated.
> 
> Also I think it would be preferable to not have the the `[innerHtml]` here as in my example, but flesh out the template more -- I just used this as an example since it's the first case I ran into.

Would "More stuff" in the above example be the "Couldn't display the template" message?  Or are you proposing that the template itself should contain at least some of the logic that steps through the notebook cells and converts them to output?

Reply:
> Would "More stuff" in the above example be the "Couldn't display the template" message? Or are you proposing that the HTML template itself should contain at least some of the logic that steps through the notebook cells and converts them to output?

I'm proposing that the HTML template itself contain more of the logic and HTML.